### PR TITLE
RV Aggregator: TakeBy (Int, Int)

### PR DESCRIPTION
**Pull request by jigold**
This is only for the (Int, Int) case. It's based on #3720, which is ready to go.

### CODE REVIEW

File: src/main/scala/is/hail/annotations/aggregators/RegionValueTakeByIntAggregator.scala [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    else {
+      if (ord.compare(p, _state.head) < 0) {
+        _state.dequeue()
+        _state += p
+      }
+    }
+  }
+
+  def seqOp(region: Region, x: Int, xm: Boolean, k: Int, km: Boolean) {
+    if (!xm && !km) {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/aggregators/RegionValueTakeByIntAggregator.scala]: I think takeBy should be like sets, the should be able to contain missing elements.  So I think this can read: `seqOp(if (xm) null else x, if (km) null else k)`.

File: src/main/scala/is/hail/annotations/aggregators/RegionValueTakeByIntAggregator.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    rvb.startArray(_state.length)
+    _state.clone.dequeueAll.toArray[(Any, Any)].map(_._1).reverse.foreach { a =>
+      rvb.addAnnotation(inputType, a)
+    }
+    rvb.endArray()
+  }
+
+  override def copy(): RegionValueTakeByIntIntAggregator = {
+    val rva = new RegionValueTakeByIntIntAggregator(n, inputType, keyType)
+    rva._state = _state.clone()
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/aggregators/RegionValueTakeByIntAggregator.scala]: Copy is meant to copy to the initial/reset/cleared state.

File: src/main/scala/is/hail/annotations/aggregators/RegionValueTakeByIntAggregator.scala [hunk 3]
```diff
@@ -0,0 +1,50 @@
+package is.hail.annotations.aggregators
+
+import is.hail.annotations._
+import is.hail.expr.types.Type
+
+import scala.collection.mutable
+
+class RegionValueTakeByIntIntAggregator(n: Int, val inputType: Type, val keyType: Type) extends RegionValueAggregator {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/aggregators/RegionValueTakeByIntAggregator.scala]: I think you can make a generic TakeBy and then have IntIntTakeBy inherit from that and just define a seqOp, calling into the generic seqOp((Any, Any)).

File: src/test/scala/is/hail/expr/ir/AggregatorsSuite.scala [hunk 4]
```diff
@@ -150,6 +150,17 @@ class AggregatorsSuite {
       args = FastIndexedSeq(I32(2)))
   }
 
+  @Test def takeBy() {
+    runAggregator(TakeBy(), TStruct("x" -> TInt32(), "y" -> TInt32()),
+      FastIndexedSeq(Row(3, 4), Row(null, null),
+      Row(null, 3), Row(21, 2), Row(11, 0), Row(45, 1)),
+      FastIndexedSeq(11, 45),
```
Comment (cseed) [inline src/test/scala/is/hail/expr/ir/AggregatorsSuite.scala]: More testing ideas: value null, also fewer elements that we're taking (as well as more).

**Comment by jigold** [comment]
Closing this in favor of a future PR.
